src/riak_kv_token_session.erl

```diff
@@ -0,0 +1,467 @@
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -------------------------------------------------------------------
+
+%% @doc Receive a session request, and attempt to gain an exclusive token for
+%% that session.  If one has been gained, allow for riak_client functions to
+%% be managed via the session
+
+-module(riak_kv_token_session).
+
+-behavior(gen_server).
+
+-include_lib("kernel/include/logger.hrl").
+
+-export(
+    [
+        session_request/4,
+        session_request_retry/1,
+        session_use/3,
+        session_release/1,
+        session_renew/1
+    ]
+).
+
+-export(
+    [
+        session_local_request/4,
+        session_local_use/4,
+        session_local_release/2,
+        session_local_renew/2
+    ]
+).
+
+-export(
+    [
+        init/1,
+        handle_call/3,
+        handle_cast/2,
+        handle_info/2,
+        terminate/2,
+        code_change/3
+    ]
+).
+
+-define(CONDITIONAL_NVAL, 5).
+-define(TOKEN_REQUEST_TIMEOUT, 12000).
+-define(TOKEN_SESSION_TIMEOUT, 30000).
+-define(TOKEN_RETRY_COUNT, 12).
+
+-record(state,
+        {
+            token_id :: token_id(),
+            token_timeout :: pos_integer(),
+            client :: riak_client:riak_client()|undefined,
+            session_id :: session_id()|undefined,
+            start_time = os:system_time(microsecond) :: pos_integer()
+        }
+    ).
+
+-type token_id() :: riak_kv_token_manager:token_id().
+-type verify_list() :: riak_kv_token_manager:verify_list().
+-type session_ref() :: binary().
+-type session_id() :: non_neg_integer().
+-type timeout_ms() :: pos_integer().
+-type erpc_error() :: {erpc, term()}.
+
+-export_type([session_ref/0]).
+
+
+%%%============================================================================
+%%% External API
+%%%============================================================================
+
+-spec session_request_retry(
+    token_id()) -> {true, session_ref()}|{false, none}|erpc_error().
+session_request_retry(TokenID) ->
+    NVal =
+        application:get_env(
+            riak_kv, stronger_conditional_nval, ?CONDITIONAL_NVAL
+        ),
+    session_request_retry(
+        TokenID,
+        NVal,
+        ?TOKEN_REQUEST_TIMEOUT,
+        ?TOKEN_SESSION_TIMEOUT,
+        ?TOKEN_RETRY_COUNT
+    ).
+
+-spec session_request_retry(
+    token_id(), 3|5, timeout_ms(), timeout_ms(), pos_integer())
+        -> {true, session_ref()}|{false, none}|erpc_error().
+session_request_retry(TokenID, NVal, RequestTimeout, TokenTimeout, Retry) ->
+    session_request_retry(
+        TokenID, NVal, RequestTimeout, TokenTimeout, Retry, 0
+    ).
+
+session_request_retry(_TokenID, _NVal, 0, _TokenTO, _Retry, _Attempts) ->
+    {false, none};
+session_request_retry(TokenID, NVal, RequestTO, TokenTO, Retry, Retry) ->
```

Comment:
1. I find counting down more logic than counting up. That is, start with Retry and then reduce until zero
2. or until 1, because now you do Retries + 1 attempts.